### ISSUE TITLE
Prefer '%s.gz' if exists and Accept-Encoding includes 'gzip'

### ISSUE THREAD

**Issue by Parakleta** [civetweb/civetweb#199]
Currently the only way to send gzipped versions of files is to ensure that the non-gzipped version doesn't exist, but this means that the non-gzipped version cannot be made available to clients that will not accept the gzip encoding.

The correct behaviour should be to send the gzipped version in preference to the non-gzipped where the encoding is accepted, and falling back to the non-gzipped if the gzipped is not found or not accepted.

**Comment by bel2125** [comment]
Good idea.

But having x.htm.gz and x.htm in the same folder (e.g. document root), and sending either x.htm.gz or x.htm as response to /x.htm will mean the server admin will have to manually put both files with the same contents in the same directory.
Still this is an enhancement as compared to the current implementation.

**Comment by Parakleta** [comment]
Unless you compress on the fly you have to have both versions available anyway.  My use cases for civetweb are either a localhost server where compression doesn't matter, or else on embedded devices where I don't have the horespower to compress on the fly but serving compressed files saves a lot of time and energy.

Putting both files is easy to achieve either using a simple script, or I am doing it as a step in a webpack build process (using [webpack/compression-webpack-plugin](http://github.com/webpack/compression-webpack-plugin)).

**Comment by mewalig** [comment]
is there any plan (or desire) to support on-the-fly compression in general-- both for send_file, as well as for streaming data sends?

**Comment by bel2125** [comment]
I did not plan to do this.
Some considerations on why not ... discussion is welcome:

* On-the-fly compression would mean there has to be a gzip algorithm in civetweb.
These algorithms either have a GPL license or a ZLIB license, but civetweb must only include MIT licensed code. While using GPL components must be absolutely avoided, calling the zlib which already ships with the operating system as an external component might be acceptable (some possible problems with the ZLIB license: see http://copyfree.org/standard/rejected).

* It makes no sense to develop an MIT licensed gzip algorithm and integrate it directly into civetweb.

* **For files:** It is simple to compress all files before starting the server. This could be done with a simple script before the server starts. It is more efficient than on-the-fly compression, because you compress every file once before the server starts, and not once again for every request.

* **For streaming data** (you probably mean *"everything except files"*): 
   * For a typical web site, most of the data is still static files: images, js, css, videos, .. and the dynamically generated contents is used to glue them together. Of course there are other applications as web pages that should be shown in a browser, and it is also possible to store the static contents on a different server and use civetweb only for dynamic contents. Still I think there is much more gain in compressing the static files as compared to compressing everything else.
   * It's much more work to implement this on-the-fly compression, since you need to buffer all the data. You probably need to know in advance how many bytes there are in zipped response (I'm not sure if there is a way to avoid this), so you probably need to buffer the entire response, which might be a lot. The current interface does not require to tell civetweb in advance how many bytes there are in an (uncompressed) response, so finding the end of a response might cause some additional problems.
   * ... as compared to compressing files more work for less effect.

* What could be useful is some scripts to create the gzip-ed files before starting the server?
Unfortunately there is no common script language between all operating systems. Civetweb could use the builtin Lua script engine for this task.
What would you think about a "before start" Lua script, that could look into the `document_root` directory, find all files, check if they are script files according to the `cgi_pattern`, `lua_script_pattern`, `lua_server_page_pattern` and `duktape_script_pattern` (scripts can not be compressed), and otherwise compress them by calling the operating system command with Lua `os.execute` (command `gzip -c image.png > image.png.gz` for Linux and `7z a -tgzip image.png image.png.gz` with 7-Zip on Windows ... not tested, just to demonstrate the idea).

**Comment by Parakleta** [comment]
I don't think there should be on-the-fly compression.  For static content it should be compressed offline before starting, there are dozens of ways to achieve this that have nothing to do with civetweb (e.g. some build tool of your choice, or a simple shell/powershell script).

For fully dynamic content I think it should generally be small enough not to matter, but where it is it could be compressed by whatever generated the content (lua or cgi or whatever).  I see no benefit to having civetweb do this work.

The only case left is hybrid content such as lua server pages, in this case maybe there could be an extension implemented where a function defined in the header or footer of the page could be passed the entire rendered page to do any final parsing (i.e. compression).  I don't know what is involved in this.

Anyway, I originally raised this issue because as it stood at the time (I haven't checked recently) civetweb doesn't even serve compressed content if it is found.

**Comment by bel2125** [comment]
I will not do anything for the hybrid contents - too much effort for too little gain.

Compressed contents works *in principle*:
For a test case: See `hello_gz.txt.gz` in https://github.com/civetweb/civetweb/tree/master/test

Compressed content support is still *not quite smoothly integrated*, e.g., 
1) a directory listing will show the `hello_gz.txt.gz` file, but you should open it with /hello_gz.txt as URL,
2) if you use `/hello_gz.txt.gz` as URL in a browser, you can download the gzip-ed file,
3) if both `hello_gz.txt` and `hello_gz.txt.gz` exist, the `hello_gz.txt` file will be preferred and `hello_gz.txt.gz` is not accessible (except for download).

I think 3) is the original issue description. 2) is probably not really an issue, 1) is a flaw that should be addressed as well.

I did not change anything in the last three months since the issue was opened. It is not a fix that requires two lines of code to be changed. I think this must be treated with a more holistic approach, and I did not have the time to do this yet. But I will take care of this later.

**Comment by jmucchiello** [comment]
gzip support could always be accomplished on the fly in the same way SSL support is done. Dynamically load a share library that will do the compression if it is configured. Then the end user can create a GPL or ZLIB based DLL that civetweb opens. No license violations as long as the end user doesn't distribute the DLL with civetweb. And even if they do, that's their problem, not yours.

**Comment by mewalig** [comment]
agree with @jmucchiello. In fact I bet that if the streaming framework were to be generalized to cover both streaming and non-streaming cases, and independent of that, to cover both compressed and uncompressed cases, then it would be easy to move all the non-streaming code into this more generalized framework and end up with about the same, maybe even lower SLOC count than we started with.

**Comment by bel2125** [comment]
The advantage of on the fly compression is of course that it will work for static files as well as for dynamic contents.
A disadvantage is, you have to compress the same file multiple times - for every request you read the uncompressed file from the disk and compress it again. If you have it already compressed, you save disk access time (you read less data) and CPU time and RAM (to calculate the compression). If network speed is not you limiting component, on the fly compression may even be slower than no compression.
So I would tend to keep the option to have pre-compressed files, even if there would be on the fly compression.

**Comment by mewalig** [comment]
yes, agree with that also: keep the option to have compressed static files that stay compressed and don't get re-compressed prior to sending, and also support the option to compress dynamic content on the fly.

It might be most flexible if the dynamic compression support (both whether it does it at all, as well as, whether the compressed content will be discarded vs saved into a cache [that may or may not be written to disk]) can be based on a settable flag in mg_request_info-- that way, the code can easily be customized to support dynamic compression on a per-request basis (if, for example, one wants to support dynamic compression for all URLs that match pattern X, but not for other URLs)-- and if anyone wanted to extend the code to do that through a configuration setting rather than a code change, it would be pretty straightforward.

**Comment by bel2125** [comment]
It's most flexible, but also the highest effort to implement, test and maintain, and probably also not that easy to understand for the user, so a good documentation is also required.
If on the fly compression is made, I would just do it for any request with a "Accept-Encoding: gzip" header set, and not store anything (if you want to have it stored, then pre-compress it).

**Comment by mewalig** [comment]
Sounds like a great start to me. Thanks

**Comment by bel2125** [comment]
I have to postpone this topic until the version 1.8 has been released.

**Comment by bel2125** [comment]
Status update: 
The code is prepared for "on the fly" compression of static files (https://github.com/civetweb/civetweb/blob/46dc39cc4fbc8f9e5b1a673dc416bdd7ad5f8106/src/civetweb.c#L8851), but this feature will not be added to the upcoming 1.10 version.

Postpone to version 1.11

**Comment by bel2125** [comment]
The current version V1.11 supports on the fly compression for all static files, if built with zlib support.

**Comment by Parakleta** [comment]
I haven't followed this in a while, but does the on-the-fly behaviour also address the issue of serving pre-compressed files? That was what this issue was originally about and you yourself expressed the advantages of supporting this in https://github.com/civetweb/civetweb/issues/199#issuecomment-167611417.

**Comment by bel2125** [comment]
Seems indeed this issue got hijacked for "implement on the fly compression" somewhere in 2016/17 - I need to check this status again in more detail.

**Comment by mewalig** [comment]
Sorry. that would have been me. The initial question seems pretty easy/straightforward to support by just checking the request Accept-Encoding header for gzip, then checking if the gzip'd version is in the file system.

**Comment by samal84** [comment]
Isn't this issue solved?

**Comment by bel2125** [comment]
It's implemented since February 2019. 
There was no "official release" and build of the new windows binaries on Sourceforge since that time, but the head version contains a stable implementation.

**Comment by bel2125** [comment]
> There was no "official release" and build of the new windows binaries on Sourceforge  ...

... now there is.
